packages/web/src/common/store/upload/sagas.ts

```diff
@@ -0,0 +1,1037 @@
+    audiusBackendInstance.getAudiusLibsTyped
+  ])
+
+  // Ensure the user is logged in
+  yield waitForAccount()
+
+  // Get the IDs ahead of time, so that stems can be associated.
+  const tracksWithIds = yield* all(
+    tracks.map((track) =>
+      call(function* () {
+        const id = yield* call([
+          audiusLibs.Track,
+          audiusLibs.Track!.generateTrackId
+        ])
+        return {
+          ...track,
+          metadata: {
+            ...track.metadata,
+            track_id: id
+          }
+        }
+      })
+    )
+  )
+
+  // Upload tracks and stems parallel together
+  const trackIds = yield* call(handleUploads, {
+    tracks: tracksWithIds,
+    kind: 'tracks'
+  })
+
+  // Get true track metadatas back
+  const newTracks = yield* call(retrieveTracks, {
+    trackIds,
+    forceRetrieveFromSource: true
+  })
+
+  // Make sure track count changes for this user
+  const account = yield* select(accountSelectors.getAccountUser)
+  yield* call(adjustUserField, {
+    user: account!,
+    fieldName: 'track_count',
+    delta: newTracks.length
+  })
+
+  // At this point, the upload was success! The rest is metrics.
+  yield* put(
+    uploadActions.uploadTracksSucceeded(
+      newTracks[0].track_id,
+      newTracks,
+      newTracks[0]
+    )
+  )
+
+  // Send analytics for any gated content
+  yield* call(
+    recordGatedTracks,
+    tracksWithIds.map((track) => track.metadata)
+  )
+
+  // If the hide remixes is turned on, send analytics event
+  for (let i = 0; i < newTracks.length; i += 1) {
+    const track = newTracks[i]
+    if (track.field_visibility?.remixes === false) {
+      yield* put(
+        make(Name.REMIX_HIDE, {
+          id: track.track_id,
+          handle: account!.handle
+        })
+      )
+    }
+  }
+
+  // Send analytics for any remixes
+  for (const remixTrack of newTracks) {
+    if (
+      remixTrack.remix_of !== null &&
+      Array.isArray(remixTrack.remix_of?.tracks) &&
+      remixTrack.remix_of.tracks.length > 0
+    ) {
+      yield* call(trackNewRemixEvent, remixTrack)
+    }
+  }
+
+  // Bust the cache so we refetch the user
+  yield* put(cacheActions.setExpired(Kind.USERS, account!.user_id))
+}
+
+function* uploadTracksAsync(
+  action: ReturnType<typeof uploadActions.uploadTracks>
+) {
+  yield waitForWrite()
+  const payload = action.payload
+  yield* put(uploadActions.uploadTracksRequested(payload))
+
+  const kind = (() => {
+    switch (payload.uploadType) {
+      case UploadType.PLAYLIST:
+        return 'playlist'
+      case UploadType.ALBUM:
+        return 'album'
+      case UploadType.INDIVIDUAL_TRACK:
+      case UploadType.INDIVIDUAL_TRACKS:
+      default:
+        return 'tracks'
+    }
+  })()
+
+  const recordEvent = make(Name.TRACK_UPLOAD_START_UPLOADING, {
+    count: payload.tracks.length,
+    kind
+  })
+  yield* put(recordEvent)
+
+  const tracks = payload.tracks
+
+  // Prep the USDC purchase conditions
+  for (const trackUpload of tracks) {
```

Comment:
This isn't from your change obv, but it feels gnarly to have this here. Could it be handled in the caller of the upload flow?